Data/models/Bookmark.swift

```diff
@@ -319,104 +265,85 @@ public class Bookmark: NSManagedObject, WebsitePresentable, Syncable {
-        let sort = orderSort ? sortRules : nil
-        let fetchRequest = NSFetchRequest<NSFetchRequestResult>()
-        fetchRequest.entity = Bookmark.entity(context: context)
-        fetchRequest.predicate =  NSPredicate(format: "syncParentDisplayUUID == %@ and isFolder == %@", searchableUUID, ignoreFolders ? "true" : "false")
-        fetchRequest.sortDescriptors = sort
-        
-        do {
-            let results = try context.fetch(fetchRequest) as? [Bookmark]
-            return results
-        } catch {
-            let fetchError = error as NSError
-            print(fetchError)
-        }
-        return nil
+        
+        let syncParentDisplayUUIDKeyPath = #keyPath(Bookmark.syncParentDisplayUUID)
+        let isFolderKeyPath = #keyPath(Bookmark.isFolder)
+                
+        let predicate = NSPredicate(format: "\(syncParentDisplayUUIDKeyPath) == %@ AND \(isFolderKeyPath) == %@",  
+                                    searchableUUID, ignoreFolders ? "YES" : "NO")
+        
+        return all(where: predicate)
     }
     
     static func get(parentSyncUUID parentUUID: [Int]?, context: NSManagedObjectContext?) -> Bookmark? {
-        guard let searchableUUID = SyncHelpers.syncDisplay(fromUUID: parentUUID) else {
+        guard let searchableUUID = SyncHelpers.syncDisplay(fromUUID: parentUUID), let context = context else {
             return nil
         }
         
-        return get(predicate: NSPredicate(format: "syncDisplayUUID == %@", searchableUUID), context: context)?.first
+        let predicate = NSPredicate(format: "syncDisplayUUID == %@", searchableUUID)
+        return first(where: predicate, context: context)
     }
     
     public static func getFolders(bookmark: Bookmark?, context: NSManagedObjectContext) -> [Bookmark] {
-    
         var predicate: NSPredicate?
         if let parent = bookmark?.parentFolder {
             predicate = NSPredicate(format: "isFolder == true and parentFolder == %@", parent)
         } else {
             predicate = NSPredicate(format: "isFolder == true and parentFolder.@count = 0")
         }
         
-        return get(predicate: predicate, context: context) ?? [Bookmark]()
+        return all(where: predicate) ?? []
     }
     
-    // TODO: Remove
     static func getAllBookmarks(context: NSManagedObjectContext) -> [Bookmark] {
-        return get(predicate: NSPredicate(format: "isFavorite == NO"), context: context) ?? [Bookmark]()
+        let predicate = NSPredicate(format: "isFavorite == NO")
+        
+        return all(where: predicate) ?? []
     }
 }
 
 // TODO: REMOVE!! This should be located in abstraction
 extension Bookmark {
     @discardableResult
     public class func remove(forUrl url: URL, save: Bool = true, context: NSManagedObjectContext) -> Bool {
-        if let bm = get(forUrl: url, context: context) as? Bookmark {
+        if let bm = get(forUrl: url, context: context)?.first {
```

Comment:
Maybe we should pass in a `count` here, so we can optimize our query, rather than just fetching the entire list and only utilizing a single one. Or alternatively, any reason this is just removing a single item? May be best to propagate and remove all?

Reply:
I will use `CRUDProtocols.first()` to get this record.

This method is used for bookmarks button toggle(the button for adding or removing bookmarks, next to history and settings buttons)

Reply:
Yeah, I just mean the implementation, we could maybe add a `get(forURL:context:count:)` and since we only care about a single record, we could allow the database to handle fetching a single record rather than doing a huge fetch and just plucking off the top.

Reply:
Looks like I forgot to fetch only 1 record for this one, thanks :)

Reply:
This LOC is outdated, it's been updated to use `CRUDProtocols.first()`